### PR TITLE
API 7.10 - `PaidMediaPurchased`

### PR DESCRIPTION
**Pull request by harshil21**
Part of #4459 

Checklist for PRs
- [x] Added ``.. versionadded:: NEXT.VERSION``, ``.. versionchanged:: NEXT.VERSION``, ``.. deprecated:: NEXT.VERSION`` or ``.. versionremoved:: NEXT.VERSION`` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)
- [x] Created new or adapted existing unit tests
- [ ] Documented code changes according to the `CSI standard <https://standards.mousepawmedia.com/en/stable/csi.html>`__
- [ ] Added myself alphabetically to ``AUTHORS.rst`` (optional)
- [x] Added new classes & modules to the docs and all suitable ``__all__`` s
- [ ] Checked the `Stability Policy <https://docs.python-telegram-bot.org/stability_policy.html>`_ in case of deprecations or changes to documented behavior

**If the PR contains API changes (otherwise, you can ignore this passage)**

- [ ] Checked the Bot API specific sections of the `Stability Policy <https://docs.python-telegram-bot.org/stability_policy.html>`_
- [ ] Created a PR to remove functionality deprecated in the previous Bot API release (`see here <https://docs.python-telegram-bot.org/en/stable/stability_policy.html#case-2>`_)

- [x] New classes:

   - [x] Added ``self._id_attrs`` and corresponding documentation
   - [x] ``__init__`` accepts ``api_kwargs`` as kw-only

- [ ]  Added new shortcuts:

   - [ ] In :class:`~telegram.Chat` & :class:`~telegram.User` for all methods that accept ``chat/user_id``
   - [ ] In :class:`~telegram.Message` for all methods that accept ``chat_id`` and ``message_id``
   - [ ] For new :class:`~telegram.Message` shortcuts: Added ``quote`` argument if methods accepts ``reply_to_message_id``
   - [ ] In :class:`~telegram.CallbackQuery` for all methods that accept either ``chat_id`` and ``message_id`` or ``inline_message_id``

- [ ]  If relevant:

   - [ ] Added new constants at :mod:`telegram.constants` and shortcuts to them as class variables
   - [x] Link new and existing constants in docstrings instead of hard-coded numbers and strings
   - [ ] Add new message types to :attr:`telegram.Message.effective_attachment`
   - [x] Added new handlers for new update types

     - [ ] Add the handlers to the warning loop in the :class:`~telegram.ext.ConversationHandler`

   - [ ] Added new filters for new message (sub)types
   - [x] Added or updated documentation for the changed class(es) and/or method(s)
   - [ ] Added the new method(s) to ``_extbot.py``
   - [ ] Added or updated ``bot_methods.rst``
   - [ ] Updated the Bot API version number in all places: ``README.rst`` (including the badge) and ``telegram.constants.BOT_API_VERSION_INFO``
   - [ ] Added logic for arbitrary callback data in :class:`telegram.ext.ExtBot` for new methods that either accept a ``reply_markup`` in some form or have a return type that is/contains :class:`~telegram.Message`

### CODE REVIEW

File: telegram/_paidmedia.py [hunk 1]
```diff
@@ -288,3 +289,52 @@ def de_json(
...
         return super().de_json(data=data, bot=bot)
+
+
+class PaidMediaPurchased(TelegramObject):
+    """This object contains information about a paid media purchase.
+
+    Objects of this class are comparable in terms of equality. Two objects of this class are
+    considered equal, if their :attr:`from_user` and :attr:`paid_media_payload` are equal.
+
+    .. versionadded:: NEXT.VERSION
```
Comment (Bibo-Joshi) [inline telegram/_paidmedia.py:300]: please move this below the note

File: telegram/_update.py [hunk 2]
```diff
@@ -383,6 +395,12 @@ class Update(TelegramObject):
     """:const:`telegram.constants.UpdateType.DELETED_BUSINESS_MESSAGES`
 
     .. versionadded:: 21.1"""
+
+    PURCHASED_PAID_MEDIA: Final[str] = constants.UpdateType.PURCHASED_PAID_MEDIA
+    """:const:`telegram.constants.UpdateType.PURCHASED_PAID_MEDIA`
+
+    .. versionadded:: NEXT.VERSION"""
```
Comment (Bibo-Joshi) [inline telegram/_update.py:402]:
```suggestion
    .. versionadded:: NEXT.VERSION
    """
```

File: telegram/ext/_handlers/paidmediapurchasedhandler.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        callback (:term:`coroutine function`): The callback function for this handler. Will be
+            called when :meth:`check_update` has determined that an update should be processed by
+            this handler. Callback signature::
+
+                async def callback(update: Update, context: CallbackContext)
+        user_id (:obj:`int` | Collection[:obj:`int`], optional): Filters requests to allow only
+            those which are from the specified user ID(s).
+
+        username (:obj:`str` | Collection[:obj:`str`], optional): Filters requests to allow only
+            those which are from the specified username(s).
```
Comment (Bibo-Joshi) [inline telegram/ext/_handlers/paidmediapurchasedhandler.py:49]: nice idea adding those directly 👍

File: telegram/ext/_handlers/paidmediapurchasedhandler.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            :obj:`bool`
+
+        """
+        if isinstance(update, Update) and update.purchased_paid_media:
+            if not self._user_ids and not self._usernames:
+                return True
+            if update.purchased_paid_media.from_user.id in self._user_ids:
+                return True
+            return update.purchased_paid_media.from_user.username in self._usernames
+        return False
```
Comment (Bibo-Joshi) [inline telegram/ext/_handlers/paidmediapurchasedhandler.py:96]:
```suggestion
        if not isinstance(update, Update) or update.purchased_paid_media:
            return False
            
        if not self._user_ids and not self._usernames:
            return True
        if update.purchased_paid_media.from_user.id in self._user_ids:
            return True
        return update.purchased_paid_media.from_user.username in self._usernames
```

yeay, early exit 🥳

File: tests/test_paidmedia.py [hunk 5]
```diff
@@ -323,3 +333,53 @@ def test_equality(self):
...
+
+    def test_de_json(self, bot):
+        json_dict = {
+            "from": self.from_user.to_dict(),
+            "paid_media_payload": self.paid_media_payload,
+        }
+        pmp = PaidMediaPurchased.de_json(json_dict, bot)
+        pmp_none = PaidMediaPurchased.de_json(None, bot)
+        assert pmp.from_user == self.from_user
+        assert pmp.paid_media_payload == self.paid_media_payload
```
Comment (Bibo-Joshi) [inline tests/test_paidmedia.py:358]:
```suggestion
        assert pmp.paid_media_payload == self.paid_media_payload
        assert pmp.api_kwargs == {}
```

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Nice PR! Just a few nitpicks :)

File: telegram/ext/_handlers/paidmediapurchasedhandler.py (hunk 3 above)
Comment (harshil21) [reply telegram/ext/_handlers/paidmediapurchasedhandler.py:49]: Ctrl C + Ctrl V 😄

**Comment by harshil21** [comment]
Should I adjust the type completness passing threshold value? It seems all the handler files have some type problems so that's why the check failed

**Comment by Bibo-Joshi** [comment]
Without having looked too closely, my guess is that there is only one typing problem. the rest propagates through the other classes b/c of the "Generic" dependencies they have on each other

**Comment by Bibo-Joshi** [comment]
apparently the initial decrease from 1.0 to 0.976 was caused by a pyright udpdate. compare the runs for

image

https://github.com/python-telegram-bot/python-telegram-bot/actions/runs/10030792499/job/27720490377
https://github.com/python-telegram-bot/python-telegram-bot/actions/runs/10082502796/job/27876898733

File: telegram/ext/_handlers/paidmediapurchasedhandler.py (hunk 4 above)
Comment (Bibo-Joshi) [reply telegram/ext/_handlers/paidmediapurchasedhandler.py:96]: this did not change yet, right? why did you resolve?

Comment (harshil21) [reply telegram/ext/_handlers/paidmediapurchasedhandler.py:96]: oh right I saw your comment thinking you were happy with the code and didn't see the suggested change lol

File: tests/test_paidmedia.py (hunk 5 above)
Comment (Bibo-Joshi) [reply tests/test_paidmedia.py:358]: looks like this fails now. did not yet check, why, though …

**Comment by Bibo-Joshi** [comment]
lgtm, then, merge at will :)

**Pull request merged**